engine/src/engine.rs

```diff
@@ -751,24 +728,45 @@ impl<'env, I: IO + Copy, E: Env, M: ModExpAlgorithm> Engine<'env, I, E, M> {
         args: &NEP141FtOnTransferArgs,
         current_account_id: &AccountId,
         handler: &mut P,
-    ) {
+    ) -> Result<SubmitResult, EngineError> {
+        const INVALID_MESSAGE: &str = "receive_erc20_tokens invalid message";
+        const UNKNOWN_NEP_141: &str = "receive_erc20_tokens unknown NEP-141";
+
         let str_amount = crate::prelude::format!("\"{}\"", args.amount);
         let output_on_fail = str_amount.as_bytes();
+        let mut local_io = self.io;
+        let engine_err = |msg: &'static str| {
+            move || {
+                sdk::log!("{}", msg);
+                local_io.return_output(output_on_fail);
+                EngineError {
+                    kind: EngineErrorKind::EvmError(ExitError::Other(Cow::Borrowed(msg))),
+                    gas_used: 0,
+                }
+            }
+        };
 
         // Parse message to determine recipient
         let mut recipient = {
             // Message format:
             //      Recipient of the transaction - 40 characters (Address in hex)
             let message = args.msg.as_bytes();
-            assert_or_finish!(message.len() >= 40, output_on_fail, self.io);
-
-            Address::new(H160(unwrap_res_or_finish!(
-                unwrap_res_or_finish!(hex::decode(&message[..40]), output_on_fail, self.io)
-                    .as_slice()
-                    .try_into(),
-                output_on_fail,
-                self.io
-            )))
+            if message.len() < 40 {
+                sdk::log!("{}", INVALID_MESSAGE);
+                self.io.return_output(output_on_fail);
+                return Err(EngineError {
+                    kind: EngineErrorKind::EvmError(ExitError::Other(Cow::Borrowed(
+                        INVALID_MESSAGE,
+                    ))),
+                    gas_used: 0,
+                });
+            }
+            let address_bytes: [u8; 20] = hex::decode(&message[..40])
+                .ok()
+                .and_then(|bytes| bytes.as_slice().try_into().ok())
+                .ok_or_else(engine_err(INVALID_MESSAGE))?;
```

Comment:
I guess `hex::decode_to_slice` will be more effective here:

```rust
let mut address_bytes = [0; 20];
hex::decode_to_slice(&message[..40], &mut address_bytes)
    .map_err(|_| engine_err(INVALID_MESSAGE))?;
```

Reply:
Good call! Refactored in https://github.com/aurora-is-near/aurora-engine/pull/869/commits/b21eb801c785eeb5ef411b283efc7f42bd8695cf